README


This archive contains the distribution TAP-DOM-Waivers,
version 0.001:

  Patching TAP::DOM, usually for test waivers

This software is copyright (c) 2014 by Steffen Schwigon.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
Revision history for TAP-DOM-Waivers

0.001     2014-03-16
        - experimental prove plugin
        - Exceptions (waivers) for TAP::DOM-like data

2011-01-10  Steffen Schwigon  <ss5@renormalist.net>

	* initial version



lib/TAP/DOM/Waivers.pm
package TAP::DOM::Waivers;
# git description: e19136e

BEGIN {
  $TAP::DOM::Waivers::AUTHORITY = 'cpan:SCHWIGON';
}
{
  $TAP::DOM::Waivers::VERSION = '0.001';
}
# ABSTRACT: Patching TAP::DOM, usually for test waivers

use 5.008;
use strict;
use warnings;

use Data::Dumper;
use Data::DPath 'dpathr';
use Clone "clone";
use Sub::Exporter -setup => {
                             exports => [ 'waive' ],
                             groups  => { all   => [ 'waive' ] },
                            };

sub waive {
        my ($dom, $waivers, $options) = @_;

        my $new_dom_ref;
        if ($options->{no_clone}) {
                $new_dom_ref = \$dom;
        } else {
                $new_dom_ref = \ (clone($dom));
        }
        foreach my $waiver (@$waivers) {
                # apply on matching dpath
                if (my @paths = @{$waiver->{match_dpath} || []}) {
                        _patch_dom_dpath( $new_dom_ref, $waiver, $_ ) foreach @paths;
                }
                elsif (my @descriptions = @{$waiver->{match_description} || []}) {
                        my @paths = map { _description_to_dpath($_) } @descriptions;
                        _patch_dom_dpath( $new_dom_ref, $waiver, $_ ) foreach @paths;
                }
        }
        return $$new_dom_ref;
}

sub _description_to_dpath {
        my ($description) = @_;

        # the '#' as delimiter is not expected in a description
        # because it has TAP semantics, however, we escape to be sure
        $description =~ s/\#/\\\#/g;

        return "//lines//description[value =~ qr#$description#]/..";
}

sub _meta_patch {
        my ($metapatch) = @_;

        my $patch;
        my $explanation;
        if ($explanation = $metapatch->{TODO}) {
                $patch = {
                          is_ok        => 1,
                          has_todo     => 1,
                          is_actual_ok => 0,
                          directive    => 'TODO',
                          explanation  => $explanation,
                         };
        } elsif ($explanation = $metapatch->{SKIP}) {
                $patch = {
                          is_ok        => 1,
                          has_skip     => 1,
                          is_actual_ok => 0,
                          directive    => 'SKIP',
                          explanation  => $explanation,
                         };
        }
        return $patch;
}

sub _patch_dom_dpath {
        my ($dom_ref, $waiver, $path) = @_;

        my $patch;
        if (exists $waiver->{metapatch}) {
                $patch = _meta_patch($waiver->{metapatch});
        } else {
                $patch = $waiver->{patch};
        }
        my $comment  = $waiver->{comment};
        my @points   = dpathr($path)->match($$dom_ref);
        foreach my $p (@points) {
                $$p->{$_} = $patch->{$_} foreach keys %$patch;
        }
}

1;

=pod

=encoding utf-8

=head1 NAME

TAP::DOM::Waivers - Patching TAP::DOM, usually for test waivers

=head1 SYNOPSIS

 use TAP::DOM;
 use TAP::DOM::Waivers 'waiver';
 
 # get TAP
 my $dom = TAP::DOM->new( tap => "somefile.tap" );
 
 # ,--------------------------------------------------------------------.
 # | Define exceptions and how to modify test results.
 # |
 # | (1) Most powerful but most complex way:
 # |     - use DPath matching and finegrained patching
 # |

 $waivers = [
             {
               # a description of what the waiver is trying to achieve
               comment     => "Force all IPv6 stuff to true",
                  
               # a DPath that matches the records to patch:
               match_dpath => [ "//lines//description[value =~ 'IPv6']/.." ],

               # apply changes to the matched records,
               # here a TODO with an explanation:
               patch       => {
                               is_ok        => 1,
                               has_todo     => 1,
                               is_actual_ok => 0,
                               explanation  => 'waiver for context xyz',
                               directive    => 'TODO',
                              },
             },
            ];
 
 # |
 # | (2) Simpler approach: 
 # |
 # |     - instead of the "patch" key above you can use "metapatches" 
 # |       for Common use-cases, like #TODO or #SKIP
 # |

 $waivers = [
             {
               comment     => "Force all IPv6 stuff to true",
               match_dpath => [ "//lines//description[value =~ 'IPv6']/.." ],
               metapatch   => { TODO => 'waiver for context xyz' },
             },
            ];

 # |
 # | (3) Even simpler:
 # |     - also provide the description as regex
 # |

 $waivers = [
             {
               comment           => "Force all IPv6 stuff to true",
               match_description => [ "IPv6" ],
               metapatch         => { TODO => 'waiver for context xyz' },
             },
            ];
 #
 # |
 # `--------------------------------------------------------------------'

 # the actual DOM patching
 my $patched_tap_dom = waiver($dom, $waivers);
 
 # do something with patched DOM
 use Data::Dumper;
 print Dumper($patched_tap_dom);
 
 # the original DOM can also be patched directly without cloning
 waiver($dom, $waivers, { no_clone => 1 });
 print Dumper(dom);
 
 # convert back to TAP from patched DOM
 print $patched_tap_dom->to_tap;
 print dom->to_tap;

=head1 NAME

TAP::DOM::Waivers - Exceptions (waivers) for TAP::DOM-like data

=head1 ABOUT

=head2 Achieve?

Test I<waivers> are exemptions to actual test results.

This module lets you ignore known issues you don't want to care about,
usually by grouping them for a certain context.

=head2 Example:

A software project might not run with IPv6 enabled but you want to see
a big SUCCESS or NO SUCCESS in an IPv4-only context, without being
disturbed by irrelevant IPv6 tests, for now.

Statically marking the problematic tests with C<#TODO> would require
to change that back and forth everytime. Dynamically marking those
tests depending on the runtime environment does not help when another
engineer actually works on fixing those IPV6 problems in the same
environment.

The solution is to create a I<waiver> which patches the IPv6 issues
away in the results B<after> you actually ran the tests, for later
evaluation.

=head2 Prove plugin

See also L<App::Prove::Plugin::Waivers|App::Prove::Plugin::Waivers>
for a way to utilze this module with B<prove> (not yet working?).

=head1 Waiver specification

=head2 How to match what to patch

This module can patch TAP-DOMs (and similar data structures, see
below) by certain criteria. The primary and most powerful way is via
Data::DPath paths, as it allows to match fuzzily against continuously
changing TAP from evolving test suites.

I use this with a big TAP database where I activate waivers as a layer
on top of TAP::DOM based evaluation. There the TAP-DOMs are just part
of a even bigger data structure, but the DPath matching still applies
there.

=head3 B<match_dpath> => [ @array_of_dpaths ]

This provides a set of dpaths that are each tried to match. The DPaths
should point to a single entry in TAP-DOM - that's why the examples
above go down into an entry to match conditions (like the
description), and then go up one level to point to the whole entry.

=head3 B<match_description> => [ @array_of_regexes ]

This is a high level frontend to I<match_dpath>. The regexes are
internally embedded in dpaths which are then used to match. The
converted internal dpaths will match fuzzy for a typical TAP-DOM
structure, in particular:

 "//lines//description[value =~ qr/$description/]/..";

Please note that this doesn't allow to specify complex conditions like
the combination of a description and a particular test success
(e.g. only the "not ok" tests with a particular description, see
examples in I<t/waivers.t>).

In combination with the also just canonically working I<metapatch>
(see below) it might create a slightly different TAP-DOM than you
expect, e.g. when you match and modify tests as '#TODO' that did not
even fail, but the metapatch marks them as 'not ok #TODO'. So the
original actual success is lost.

It might be still "quite ok" and worth the less complexity but
consider using I<match_dpath> for better control.

=head2 Patch specs

=head3 B<patch> => { %patch_spec }

A hash entry key B<patch> contains single keys that overwrite
respective fields of a TAP-DOM entry.

This allows finegrained control but it's somewhat difficult if you are
not familiar with the details of how a TAP situation looks like in a
TAP-DOM.

Therefore you can describe more abstract use-cases with
I<metapatches>.

=head3 B<metapatch> => { %patch_spec }

A key B<metapatch> declares a common use-case. Inside a metapatch the
key describes the use case (like 'TODO'), and the value is the most
significant thingie (eg. the explanation).

Currently these metapatches are supported:

=over 4

=item * B<TODO> => I<explanation>

=item * B<SKIP> => I<explanation>

=back

When such a metapatch is found it is converted internally into an
equivalent detailed patch, as described above.

=head2 Comments

The key B<comment> is not strictly needed. It will help once there is
some logging.

=head1 Back from DOM to TAP

Usually you regenerate a semantically comparable TAP document from the
DOM via L<TAP::DOM::to_tap|TAP::DOM/to_tap>.

=head1 API

=head2 waive ($dom, $waivers, $options)

This applies a set of waivers to a TAP-DOM.

The C<TAP-DOM> is usually a real L<TAP::DOM|TAP::DOM> but don't have
to. It is explicitely allowed to provide similar data structures,
e.g., bigger structures that only contain TAP-DOMs in sub
structures. It's your responsibility to provide something meaningful.

If you match with C<match_dpath> you have control whether to use the
surrounding data structures to match or not.

If a waiver does not match, nothing happens.

=head1 AUTHOR

Steffen Schwigon, C<< <ss5 at renormalist.net> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-tap-dom-waivers at
rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=TAP-DOM-Waivers>.  I
will be notified, and then you'll automatically be notified of
progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc TAP::DOM::Waivers

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=TAP-DOM-Waivers>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/TAP-DOM-Waivers>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/TAP-DOM-Waivers>

=item * Search CPAN

L<http://search.cpan.org/dist/TAP-DOM-Waivers/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright 2011 Steffen Schwigon.

This program is free software; you can redistribute it and/or modify
it under the terms of either: the GNU General Public License as
published by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=head1 AUTHOR

Steffen Schwigon <ss5@renormalist.net>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Steffen Schwigon.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut

__END__


# Idea brain dump

# Other match criteria, eg. by line:

#  my $waivers = [
#                 {
#                   match_lines => [ 7, 9, 15 ],
#                   patch       => { ... },
#                 }
#                ];

# By number:

#  my $waivers = [
#                 {
#                   match_numbers => [ 5, 7, 12 ],
#                   patch         => { ... },
#                 }
#                ];

# By descriptions:

#  my $waivers = [
#                 {
#                   match_descriptions => [ 'IPv6' ],
#                   patch              => { ... },
#                 }
#                ];


lib/TAP/DOM/Waivers/Formatter.pm
package TAP::DOM::Waivers::Formatter;
BEGIN {
  $TAP::DOM::Waivers::Formatter::AUTHORITY = 'cpan:SCHWIGON';
}
{
  $TAP::DOM::Waivers::Formatter::VERSION = '0.001';
} # better TAP::Formatter::DOM?
# ABSTRACT: (incomplete) 'prove' plugin support for TAP::DOM::Waivers

use strict;
use warnings;

use TAP::DOM::Waivers::Formatter::Session; # better TAP::Formatter::DOM::Session?
use Data::Dumper;

use base qw( TAP::Base );
use accessors qw(
                        verbosity
                        tests
                        session_class
                        escape_output
                        stdout
                        sessions
                        waiver
               );

our @TFW_ARGS = qw(
                          waiver
                          escape_output
                 );

# global variables
sub _initialize {
    my ($self, $args) = @_;

    $args ||= {};
    $self->SUPER::_initialize($args);

    my $stdout_fh = IO::File->new_from_fd( fileno(STDOUT), 'w' )
      or die "Error opening STDOUT for writing: $!";

    $self->verbosity( 0 )
          ->stdout( $stdout_fh )
          ->escape_output( 0 )
          ->session_class( "TAP::DOM::Waivers::Formatter::Session" )
          ->sessions( [] )
          ;

    # Laziness...
    # trust the user knows what they're doing with the args:
    foreach my $key (keys %$args) {
            $self->info( "INIT args.$key = ", $args->{$key}, "\n") if $self->can( $key );
            $self->$key( $args->{$key} ) if ($self->can( $key ));
    }
    $self->info( "KEY: $_ = ".$args->{$_} ) foreach keys %$args;

    $self->check_for_overrides_in_env;

    $self->info( "WAIVER FILE: ".$self->waiver );


    return $self;
}

sub check_for_overrides_in_env {
    my $self = shift;

    foreach my $arg (@TFW_ARGS) {
            my $val = $ENV{"TAP_FORMATTER_WAIVERS_".uc($arg)};
            $self->info( "arg: $arg = $val" ) if defined $val;
            $self->$arg( $val ) if defined $val;
    }

    return $self;
}

# Called by Test::Harness before any test output is generated.
sub prepare {
    my ($self, @tests) = @_;
    # warn ref($self) . "->prepare called with args:\n" . Dumper( \@tests );
    $self->info( 'PREPARE ', scalar @tests, ' tests' );
    $self->tests( [@tests] );
}

sub open_test {
    my ($self, $test, $parser) = @_;
    #warn ref($self) . "->open_test called with args: " . Dumper( [$test, $parser] );
    $self->info( 'OPEN_TEST');
    my $session = $self->session_class->new({ test => $test,
                                              parser => $parser,
                                              formatter => $self,
                                            });
    push @{ $self->sessions }, $session;
    return $session;
}

sub summary {
        my ($self, $aggregate) = @_;

        if (! $self->silent) {
                $self->info("SUMMARY");
        }

        # $self->info( "PARSER:   ", Dumper($self->sessions->[0]->parser) );
        # $self->info( "SESSIONS: ", Dumper($self->sessions) );

        return $self;
}

sub log {
    my $self = shift;
    push @_, "\n" unless grep {/\n/} @_;
    $self->_output( @_ );
    return $self;
}

sub info {
    my $self = shift;
    return unless $self->verbose;
    return $self->log( @_ );
}

sub log_test {
    my $self = shift;
    return if $self->really_quiet;
    return $self->log( @_ );
}

sub log_test_info {
    my $self = shift;
    return if $self->quiet;
    return $self->log( @_ );
}

sub _output {
    my $self = shift;
    return if $self->silent;
    if (ref($_[0]) && ref( $_[0]) eq 'SCALAR') {
        # DEPRECATED: printing HTML:
        print { $self->stdout } ${ $_[0] };
    } else {
        unshift @_, '# ' if $self->escape_output;
        print { $self->stdout } @_;
    }
}

sub verbose {
    my $self = shift;
    # emulate a classic accessor for compat w/TAP::Formatter::Console:
    if (@_) { $self->verbosity(1) }
    return $self->verbosity >= 1;
}

sub quiet {
    my $self = shift;
    # emulate a classic accessor for compat w/TAP::Formatter::Console:
    if (@_) { $self->verbosity(-1) }
    return $self->verbosity <= -1;
}

sub really_quiet {
    my $self = shift;
    # emulate a classic accessor for compat w/TAP::Formatter::Console:
    if (@_) { $self->verbosity(-2) }
    return $self->verbosity <= -2;
}

sub silent {
    my $self = shift;
    # emulate a classic accessor for compat w/TAP::Formatter::Console:
    if (@_) { $self->verbosity(-3) }
    return $self->verbosity <= -3;
}

1;

__END__

=pod

=encoding utf-8

=head1 NAME

TAP::DOM::Waivers::Formatter - (incomplete) 'prove' plugin support for TAP::DOM::Waivers

=head2 METHODS

=head3 check_for_overrides_in_env

=head3 info

=head3 log

=head3 log_test

=head3 log_test_info

=head3 open_test

=head3 prepare

=head3 quiet

=head3 really_quiet

=head3 silent

=head3 summary

=head3 verbose

=head1 AUTHOR

Steffen Schwigon <ss5@renormalist.net>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Steffen Schwigon.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/TAP/DOM/Waivers/Formatter/Session.pm
package TAP::DOM::Waivers::Formatter::Session;
BEGIN {
  $TAP::DOM::Waivers::Formatter::Session::AUTHORITY = 'cpan:SCHWIGON';
}
{
  $TAP::DOM::Waivers::Formatter::Session::VERSION = '0.001';
} # better TAP::Formatter::DOM::Session?
# ABSTRACT: (incomplete) 'prove' plugin support for TAP::DOM::Waivers

use strict;
use warnings;

use base qw( TAP::Base );
use accessors qw( test formatter parser results html_id meta closed taplines );

use YAML::Any;
use TAP::DOM;
use TAP::DOM::Waivers 'waive';
use File::Temp 'tempfile';

sub _slurp {
        my ($filename) = @_;

        local $/;
        open (my $F, "<", $filename) or die "Cannot read $filename";
        return <$F>;
}

sub _initialize {
    my ($self, $args) = @_;

    $args ||= {};
    $self->SUPER::_initialize($args);

    $self
     ->results([])
     ->meta({})
     ->closed(0)
     ->taplines([])
      ;
    foreach my $arg (qw( test parser formatter )) {
            $self->$arg($args->{$arg}) if defined $args->{$arg};
    }

    $self->info( "INITIALIZE: " . $self->test . ':' );

    return $self;
}

sub result {
    my ($self, $result) = @_;

    push @{$self->taplines}, $result->raw;
}

# Called by TAP::?? to indicate there are no more test results coming
sub close_test {
    my ($self, @args) = @_;

    $self->info( "CLOSE" );

    $self->info( $_ ) foreach @{$self->taplines};

    my ($FH_tmptap, $tmptap) = tempfile();
    my $waiverfile     = $self->formatter->waiver;
    my $tapfile        = $self->test;
    my $waivers        = YAML::Any::Load(_slurp($waiverfile));
    my $tapdom         = TAP::DOM->new(tap => _slurp($tapfile));
    my $patched_tapdom = waive($tapdom, $waivers);
    $self->info( "*** PATCHED_TAP: ".$patched_tapdom->to_tap );

    $self->closed(1);
    return;
}

sub as_report {
    my ($self) = @_;
    my $p = $self->parser;
    my $r = {
        test => $self->test,
        results => $self->results,
        AFFE => "ZOMTEC",
    };

    $self->info( "AS_REPORT" );

    return $r;
}

sub log {
    my ($self, @args) = @_;
    $self->formatter->log_test(@args);
}

sub info {
    my ($self, @args) = @_;
    $self->formatter->log_test_info(@args);
}

1;

__END__

=pod

=encoding utf-8

=head1 NAME

TAP::DOM::Waivers::Formatter::Session - (incomplete) 'prove' plugin support for TAP::DOM::Waivers

=head2 METHODS

=head3 as_report

=head3 close_test

=head3 info

=head3 log

=head3 result

=head1 AUTHOR

Steffen Schwigon <ss5@renormalist.net>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Steffen Schwigon.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

use strict;
use warnings;

use 5.008;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Patching TAP::DOM, usually for test waivers",
  "AUTHOR" => "Steffen Schwigon <ss5\@renormalist.net>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "TAP-DOM-Waivers",
  "EXE_FILES" => [
    "bin/tap-waive"
  ],
  "LICENSE" => "perl",
  "NAME" => "TAP::DOM::Waivers",
  "PREREQ_PM" => {
    "Clone" => 0,
    "Data::DPath" => "0.42",
    "Data::Dumper" => 0,
    "File::Temp" => 0,
    "Sub::Exporter" => 0,
    "TAP::Base" => 0,
    "TAP::DOM" => "0.06",
    "Test::Harness" => "3.22",
    "Test::More" => 0,
    "YAML::Any" => 0,
    "accessors" => 0,
    "base" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {},
  "VERSION" => "0.001",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



